include/mra/misc/convolutiondata.h

```diff
@@ -0,0 +1,213 @@
+        cachemutex.lock();
+        auto it = rnlijcache.find(key);
+        cachemutex.unlock();
+        if (it != rnlijcache.end()) {
+          const auto& r = it->second;
+          return r;
+        }
+        Tensor<T, 1> R(4*K);
+        Tensor<T, 2> rnlij(K, K);
+        auto R_view = R.current_view();
+
+        const auto& rnlp1 = make_rnlp(n, lx-1);
+        const auto& rnlp2 = make_rnlp(n, lx);
+        auto rnlp1_view = rnlp1.current_view();
+        auto rnlp2_view = rnlp2.current_view();
+
+        std::array<Slice,1> slice1 = {Slice(0, 2*K)};
+        R_view(slice1) = rnlp1_view(slice1);
+        std::array<Slice,1> slice2 = {Slice(2*K, 4*K)};
+        R_view(slice2) = rnlp2_view(slice1);
+
+        T scale = std::pow(T(0.5), T(0.5*n));
+        R_view *= scale;
+        auto rnlij_view = rnlij.current_view();
+        detail::inner(c.current_view(), R_view, rnlij_view);
+
+        cachemutex.lock();
+        if (rnlijcache.find(key) == rnlijcache.end()) {
+          assert(rnlijcache.find(key) == rnlijcache.end());
+          rnlijcache.emplace(std::move(key), std::move(rnlij));
+        }
+
+        it = rnlijcache.find(key);
+        cachemutex.unlock();
+        const auto& r = it->second;
+        return r;
+      }
+
+      const ConvolutionData<T>& make_nonstandard (const Level n, const Translation lx) {
+        mra::Key<NDIM> key(n, std::array<Translation, NDIM>({lx}));
+        auto it = nscache.find(key);
+        if (it != nscache.end()) {
+          return std::move(it->second);
+        }
+
+        Tensor<T, 2> tmp(2*K, 2*K);
+        const Tensor<T, 2>& rm = make_rnlij(n+1, 2*lx-1);
+        const Tensor<T, 2>& r0 = make_rnlij(n+1, 2*lx);
+        const Tensor<T, 2>& rp = make_rnlij(n+1, 2*lx+1);
+
+        auto tmp_view = tmp.current_view();
+        auto rm_view = rm.current_view();
+        auto r0_view = r0.current_view();
+        auto rp_view = rp.current_view();
+
+        std::array<Slice,2> slice = {Slice(0, K), Slice(0, K)};
+        tmp_view(slice) = r0_view;
+        slice = {Slice(0, K), Slice(K, 2*K)};
+        tmp_view(slice) = rm_view;
+        slice = {Slice(K, 2*K), Slice(0, K)};
+        tmp_view(slice) = rp_view;
+        slice = {Slice(K, 2*K), Slice(K, 2*K)};
+        tmp_view(slice) = r0_view;
+
+        const auto& hgT = functiondata.get_hgT();
+        auto hgT_view = hgT.current_view();
+        Tensor<T, 2> R(2*K, 2*K), work(2*K, 2*K);
+        auto R_view = R.current_view();
+        transform(tmp_view, hgT_view, R_view, work.data());
+        Tensor<T, 2> S(K, K);
+        auto S_view = S.current_view();
+        slice = {Slice(0, K), Slice(0, K)};
+        S_view(slice) = R_view(slice);
+
+        auto obj = ConvolutionData<T>();
+        obj.R = std::move(R);
+        obj.S = std::move(S);
+
+        cachemutex.lock();
+        if (nscache.find(key) == nscache.end()) {
+          assert(nscache.find(key) == nscache.end());
+          nscache.emplace(std::move(key), std::move(obj));
```

Comment:
Don't move the key here, you're using it below.

Reply:
Addressed in recent commit.